src/poetry/core/utils/helpers.py

```diff
@@ -91,6 +92,27 @@ def safe_rmtree(path: str | Path) -> None:
     shutil.rmtree(path, onerror=_on_rm_error)
 
 
+def robust_rmtree(path: str, max_timeout: float = 1) -> None:
+    """
+    Robustly tries to delete paths.
+    Retries several times if an OSError occurs.
+    If the final attempt fails, the Exception is propagated
+    to the caller.
+    """
+    timeout = 0.001
+    while timeout < max_timeout:
+        try:
+            shutil.rmtree(path)
+            return  # Only hits this on success
+        except OSError:
+            # Increase the timeout and try again
+            time.sleep(timeout)
+            timeout *= 2
+
+    # Final attempt, pass any Exceptions up to caller.
+    safe_rmtree(path)
```

Comment:
Wouldn't we want to start with the is_symlink logic from `safe_rmtree` itself?

If we do that, we can remove the `safe_rmtree` function entirely as this is the only place it's used. 

```suggestion
    if Path(path).is_symlink():
        return os.unlink(str(path))

    timeout = 0.001
    while timeout < max_timeout:
        try:
            shutil.rmtree(path)
            return  # Only hits this on success
        except OSError:
            # Increase the timeout and try again
            time.sleep(timeout)
            timeout *= 2

    # Final attempt, pass any Exceptions up to caller.
     shutil.rmtree(path, onerror=_on_rm_error)
```

Reply:
I'm happy with any solution as long as it's working on Windows.

Reply:
I've taken the code from @cdgriffith, but with a small change.

From what I can see `os.unlink` can also throw exceptions on Windows if the resource is in use, so I added the check and call for this in the big while loop to catch this scenario as well.